x-pack/filebeat/input/cometd/input_test.go

```diff
@@ -0,0 +1,374 @@
+
+		outlet := &mockedOutleter{
+			onEventHandler: func(event beat.Event) bool {
+				eventsCh <- event
+				return true
+			},
+		}
+		connector := &mockedConnector{
+			outlet: outlet,
+		}
+
+		var expected bay.MaybeMsg
+		expected.Msg.Data.Event.ReplayID = 1234
+		expected.Msg.Data.Payload = []byte(`{"CountryIso": "IN"}`)
+		expected.Msg.Channel = firstChannel
+
+		config1 := map[string]interface{}{
+			"channel_name":              firstChannel,
+			"auth.oauth2.client.id":     "client.id",
+			"auth.oauth2.client.secret": "client.secret",
+			"auth.oauth2.user":          "user",
+			"auth.oauth2.password":      "password",
+		}
+		config2 := map[string]interface{}{
+			"channel_name":              secondChannel,
+			"auth.oauth2.client.id":     "client.id",
+			"auth.oauth2.client.secret": "client.secret",
+			"auth.oauth2.user":          "user",
+			"auth.oauth2.password":      "password",
+		}
+
+		// create Server
+		r := http.HandlerFunc(oauth2Handler)
+		server := httptest.NewServer(r)
+		serverURL = server.URL
+		config1["auth.oauth2.token_url"] = serverURL + "/token"
+		config2["auth.oauth2.token_url"] = serverURL + "/token"
+
+		// get common config
+		cfg1 := common.MustNewConfigFrom(config1)
+		cfg2 := common.MustNewConfigFrom(config2)
+
+		var inputContext finput.Context
+
+		// initialize inputs
+		input1, err := NewInput(cfg1, connector, inputContext)
+		require.NoError(t, err)
+		require.NotNil(t, input1)
+
+		input2, err := NewInput(cfg2, connector, inputContext)
+		require.NoError(t, err)
+		require.NotNil(t, input2)
+
+		// run input
+		input1.Run()
+		defer input1.Stop()
+
+		event1 := <-eventsCh
+		assertEventMatches(t, expected, event1)
+
+		// run input
+		input2.Run()
+		defer input2.Stop()
+
+		event2 := <-eventsCh
+		assertEventMatches(t, expected, event2)
+
+		// close server
+		server.Close()
+	})
+
+	t.Run("Check gracefully shutdown input", func(t *testing.T) {
+		eventsCh := make(chan beat.Event)
+
+		outlet := &mockedOutleter{
+			onEventHandler: func(event beat.Event) bool {
+				eventsCh <- event
+				return true
+			},
+		}
+		connector := &mockedConnector{
+			outlet: outlet,
+		}
+		var inputContext finput.Context
+
+		var msg bay.MaybeMsg
+		msg.Msg.Data.Event.ReplayID = 1234
+		msg.Msg.Data.Payload = []byte(`{"CountryIso": "IN"}`)
+		msg.Msg.Channel = firstChannel
+
+		config := map[string]interface{}{
+			"channel_name":              firstChannel,
+			"auth.oauth2.client.id":     "client.id",
+			"auth.oauth2.client.secret": "client.secret",
+			"auth.oauth2.user":          "user",
+			"auth.oauth2.password":      "password",
+		}
+
+		r := http.HandlerFunc(oauth2Handler)
+		server := httptest.NewServer(r)
+		serverURL = server.URL
+		config["auth.oauth2.token_url"] = serverURL + "/token"
+
+		cfg := common.MustNewConfigFrom(config)
+
+		input, err := NewInput(cfg, connector, inputContext)
+		require.NoError(t, err)
+		require.NotNil(t, input)
+
+		// run input
+		input.Run()
+
+		go func() {
```

Comment:
If you comment out this entire goroutine the test still passes which isn't right.

Reply:
Ok there are multiple bugs here, several on the Bayeux side.

1) The `eventsCh` here is only ever used from this test itself. The way the test is written we receive a single event into a slice of `beat.Event` and then never look at the channel again. If you wanted to test that `eventsCh` is closed you would want `for for range eventsCh` instead of `for range []beat.Event{<-eventsCh}`. Not that `eventsCh` will never be closed after calling `Wait` though, because it is only ever used in this test.

2) The message channel needs a buffer size of 1:

```diff
diff --git a/x-pack/filebeat/input/cometd/input.go b/x-pack/filebeat/input/cometd/input.go
index a8aaccb139..9ee31ca5d0 100644
--- a/x-pack/filebeat/input/cometd/input.go
+++ b/x-pack/filebeat/input/cometd/input.go
@@ -134,7 +134,7 @@ func NewInput(
        }

        // Creating a new channel for cometd input.
-       in.msgCh = make(chan bay.MaybeMsg)
+       in.msgCh = make(chan bay.MaybeMsg, 1)

        // Build outlet for events.
        in.outlet, err = connector.Connect(cfg)
```

Otherwise if the context is cancelled before the `bayeux.Channel` call executes then the `out <- MaybeMsg{Err: err}` call below will block forever waiting for a reader that isn't there anymore. The `out` channel is also never closed in this case.

```go
func (b *Bayeux) Channel(ctx context.Context, out chan MaybeMsg, r string, creds Credentials, channel string) chan MaybeMsg {
	b.creds = creds
	err := b.getClientID(ctx)
	if err != nil {
		out <- MaybeMsg{Err: err} // FIX: This will block until the reader receives from out
	}
	b.subscribe(ctx, channel, r) // FIX: The error from subscribe is never checked
	c := b.connect(ctx, out)
	return c
}
```

Reply:
To test that the channel gets closed inside bayeux, you need some way to know when the `func (in *cometdInput) run() error` internal function has exited. You currently have no way to do this from a test.

Reply:
1. Right.
2. Got it. Will update it.